src/include/migraphx/matcher.hpp

```diff
@@ -895,9 +895,9 @@ inline auto has_value(T x, std::size_t atol_mult = 10, std::size_t rtol_mult = 1
             }
             else
             {
-                auto eps = std::numeric_limits<type>::epsilon();
+                float eps = std::numeric_limits<type>::epsilon();
                 if(std::all_of(v.begin(), v.end(), [&](auto val) {
-                       return std::fabs(val - static_cast<type>(x)) <
+                       return std::fabs(static_cast<float>(val) - static_cast<float>(x)) <
                               eps * (atol_mult + rtol_mult * std::fabs(val));
```

Comment:
Static casting to a float is problematic. I specifically had it cast to the type of the literal because of how fp16 cannot represent all of the values fp32 can. This leads to problems with constants that are already quantized to fp16 in pre-quantized models.

Reply:
Casting to float is problematic because it cant represent values in `double`. 

> Ex: model quantized a value x to fp16 which becomes a value y. If you upcast value y to fp32, it will stay as value y when comparing to the target value.

I dont really understand what the issue is here.

Reply:
https://github.com/ROCm/AMDMIGraphX/blob/432e4e526a9fe9bf73af57b8f2148c9a0f0e20d0/test/rewrite_gelu_test.cpp#L127

There's an example of the `fit_const` value being slightly off in the prequantized model. It's exactly the same value in fp16 however.

Reply:
Reverted change

Reply:
> Ex: model quantized a value x to fp16 which becomes a value y. If you upcast value y to fp32, it will stay as value y when comparing to the target value. 

I don't understand.  It is doing comparing like `a < b` where `b` would i suppose be 64-bit floating point value because atol and rtol are std::size_t 

now `a` is `val -x`. It doesn't matter because it is upcasting.

Reply:
> now `a` is `val -x`. It doesn't matter because it is upcasting.
It can matter. In the example I linked to this number in fp16: `0.044708251953125` is already the closest representation of `0.044715f`, which is what we're searching for. It's not "close enough" it's exactly the right representation.